src/unistd.rs

```diff
@@ -2136,3 +2146,368 @@ mod setres {
+                },
+            }
+
+        };
+
+        match i {
+            0 => {
+                if !res.is_null() {
+                    unsafe { Some(Ok(Group::from(*res))) }
+                } else {
+                    None
+                }
+            },
+            _ => { Some(Err(Error::Sys(Errno::last()))) },
+        }
+    }
+}
+
+#[cfg(not(any(target_os = "android", 
+              target_os = "ios", 
+              target_os = "macos", 
+              target_env = "musl",
+              target_arch = "mips",
+              target_arch = "mips64",
+              target_arch = "mips64el",
+              target_arch = "mipsel",
+              target_arch = "s390x")))] 
+mod usergroupiter {
+    use libc;
+    use libc::{c_char};
+    use Result;
+    use errno::Errno;
+    use super::{Error, User, Group, PWGRP_BUFSIZE};
+    use std::{mem, ptr};
+
+    /// This iterator can be used to get all of the users on the system. For example:
+    ///
+    /// ```
+    /// use nix::unistd::Users;
+    /// eprintln!("Users on this system:\nname\tuid");
+    /// Users::new()
+    ///     .map(|e|e.map(
+    ///         |pw| println!("{}\t{}",
+    ///                       pw.name,
+    ///                       pw.uid)))
+    ///     .collect::<Vec<_>>();
+    /// 
+    /// ```
+    ///
+    /// This iterator should not be used in different threads without synchronization; while doing so
+    /// will not cause undefined behavior, because modern systems lack re-entrant versions of
+    /// `setpwent` and `endpwent`, it is very likely that iterators running in different threads will
+    /// yield different numbers of items.
+    #[derive(Debug)]
+    pub struct Users(pub usize);
+
+    impl Users {
+        pub fn new() -> Self {
+            unsafe { libc::setpwent(); } 
+            Users(PWGRP_BUFSIZE)
+        }
+
+        pub fn with_bufsize(bufsize: usize) -> Self {
+            unsafe { libc::setpwent(); }
+            Users(bufsize)
+        }
+    }
+
+    impl Iterator for Users {
+        type Item = Result<User>;
+        fn next(&mut self) -> Option<Result<User>> {
+            
+            let mut cbuf = vec![0 as c_char; self.0];
+            let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+            let mut res = ptr::null_mut();
+            
+            let i = unsafe { 
+                Errno::clear();
+                libc::getpwent_r(&mut pwd, cbuf.as_mut_ptr(), self.0, &mut res) 
+            };
+
+            match i {
+                0 if !res.is_null() => {
+                    unsafe { Some(Ok(User::from(*res))) }
+                },
+                libc::ERANGE => { Some(Err(Error::Sys(Errno::last()))) },
+                _ => None
+            }
+        }
+    }
+
+    impl Drop for Users {
+        fn drop(&mut self) {
+            unsafe { libc::endpwent() };
+        }
+    }
+
+    /// This iterator can be used to get all of the groups on the system. For example:
+    ///
+    /// ```
```

Comment:
Please put this in an `# Examples` section.

Reply:
This seems done